cosign/verifier.go

```diff
@@ -0,0 +1,324 @@
+		}
+		opts.RekorClient = rekorClient
+
+		// This performs an online fetch of the Rekor public keys, but this is needed
+		// for verifying tlog entries (both online and offline).
+		opts.RekorPubKeys, err = cosign.GetRekorPubs(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("getting Rekor public keys: %w", err)
+		}
+	}
+	return opts, nil
+}
+
+// Name returns the name of the verifier.
+func (v *Verifier) Name() string {
+	return v.name
+}
+
+// Type returns the type of the verifier which is always `cosign`.
+func (v *Verifier) Type() string {
+	return cosignVerifierType
+}
+
+// Verifiable returns true if the artifact is a cosign signature.
+func (v *Verifier) Verifiable(artifact ocispec.Descriptor) bool {
+	return artifact.ArtifactType == artifactTypeCosign &&
+		artifact.MediaType == ocispec.MediaTypeImageManifest
+}
+
+// Verify verifies the cosign signature.
+func (v *Verifier) Verify(ctx context.Context, opts *ratify.VerifyOptions) (*ratify.VerificationResult, error) {
+	v.MapSigVerifier(ctx, opts)
+
+	subjectRef, err := registry.ParseReference(opts.Subject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse subject reference: %w", err)
+	}
+
+	signatureLayers, err := getSignatureBlobDesc(ctx, opts.Store, subjectRef, opts.SubjectDescriptor)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get signature blob descriptor: %w", err)
+	}
+
+	numResults := len(signatureLayers)
+	if numResults == 0 {
+		return nil, fmt.Errorf("unable to locate reference with artifactType %s", artifactTypeCosign)
+	}
+
+	signatureDesc := signatureLayers[numResults-1]
+
+	staticOpts, err := staticLayerOpts(signatureDesc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Cosign signature  %w", err)
+	}
+	signatureBlob, err := opts.Store.FetchBlobContent(ctx, subjectRef.Repository, signatureDesc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch signature blob: %w", err)
+	}
+	sig, err := static.NewSignature(signatureBlob, signatureDesc.Annotations[static.SignatureAnnotationKey], staticOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate the cosign signature: %w", err)
+	}
+
+	// create the hash of the subject image descriptor (used as the hashed payload)
+	signatureDescHash := v1.Hash{
+		Algorithm: signatureDesc.Digest.Algorithm().String(),
+		Hex:       signatureDesc.Digest.Hex(),
+	}
+	result := &ratify.VerificationResult{
+		Verifier: v,
+	}
+	_, err = cosign.VerifyImageSignature(ctx, sig, signatureDescHash, v.CheckOpts)
+	if err != nil {
+		result.Err = err
+		return result, nil
+	}
+
+	result.Description = "cosign signature verification succeeded"
+	return result, nil
+}
+
+// MapSigVerifier maps and returns a signature verifier based on the provided VerifyCommand and CheckOpts.
+// It supports different types of verifiers including key references, security keys, and certificate references.
+func (v *Verifier) MapSigVerifier(ctx context.Context, opts *ratify.VerifyOptions) (err error) {
```

Comment:
one concern: seems `MapSigVerifier` gets executed per each artifact validation request, so the verifier's options will be updated in each request, how do we ensure its goroutine-safe?

Reply:
I would protect access v.CheckOpts, thus avoid shared state